src/webgpu/api/operation/rendering/depth.spec.ts

```diff
@@ -3,13 +3,211 @@ Test related to depth buffer, depth op, compare func, etc.
 `;
 
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { TypedArrayBufferView } from '../../../../common/util/util.js';
 import { kDepthStencilFormats, kTextureFormatInfo } from '../../../capability_info.js';
 import { GPUTest } from '../../../gpu_test.js';
 
+class DepthTest extends GPUTest {
+  checkStencilCompareFunction(
+    compareFunction: GPUCompareFunction,
+    stencilColor: number[],
+    stencilRefValue: number,
+    stencilColorExpected: boolean
+  ) {
+    const depthSpencilFormat: GPUTextureFormat = 'depth24plus-stencil8';
+    const baseColor = [255, 255, 255, 255];
+
+    const baseStencilState = {
+      compare: 'always' as GPUCompareFunction,
+      failOp: 'keep' as GPUStencilOperation,
+      depthFailOp: 'keep' as GPUStencilOperation,
+      passOp: 'replace' as GPUStencilOperation,
+    };
+    const stencilState = {
+      compare: compareFunction,
+      failOp: 'keep' as GPUStencilOperation,
+      depthFailOp: 'keep' as GPUStencilOperation,
+      passOp: 'keep' as GPUStencilOperation,
+    };
+
+    const baseState = {
+      format: depthSpencilFormat,
+      depthWriteEnabled: false,
+      depthCompare: 'always' as GPUCompareFunction,
+      stencilFront: baseStencilState,
+      stencilBack: baseStencilState,
+    };
+
+    const state = {
+      format: depthSpencilFormat,
+      depthWriteEnabled: false,
+      depthCompare: 'always' as GPUCompareFunction,
+      stencilFront: stencilState,
+      stencilBack: stencilState,
+    };
+
+    this.runStencilStateTest(
+      baseState,
+      state,
+      stencilRefValue,
+      stencilColorExpected ? stencilColor : baseColor
+    );
+  }
+
+  runStencilStateTest(
+    baseState: GPUDepthStencilState,
+    state: GPUDepthStencilState,
+    stencilRefValue: number,
+    stencilColor: number[]
```

Comment:
I think `stencilColor` is mixed up with an `expectedColor`. If `stencilColorExpected` is `false` then this `stencilColor` will just be `baseColor` and then both draw calls below might draw the same color and you can't tell whether the stencil test passed or not.

Reply:
I thought that the result could be different because of the 'stencil state'. `baseState` has `always` compare function with 'replace' passOp. On the other hand, 'state' has the given compare function with 'keep' passOp. Can't the state difference affect the test result? I referred to the `CheckStencilCompareFunction`.
[1] https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/tests/end2end/DepthStencilStateTests.cpp;l=192;drc=3874c9832e2de7ebf55eb4cad2bf9683556fb5e9?q=depthStencilState&ss=chromium

Reply:
The problem is this `stencilColor` variable is actually getting set to `expectStencilColor ? kStencilColor : kBaseColor`. But then it's used as the draw color for the second draw call. If expectStencilColor is false, then both draw calls draw kBaseColor, and we expect kBaseColor, so the test passes regardless of which draw call's results we see.

This is my understanding of what this code means to do: https://github.com/kainino0x/cts/commit/c502adcb37a5d17c45f8f404b48eca4c4442f593
Or maybe I've misunderstood the intent of the original test? Since I didn't understand why it had so many colors, either.

Reply:
Ah, I see. I thought that the color could be affected by the stencil reference value and the stencil face state. But, it might be strange to draw the same color twice in the case that expecteStencilColor is false. I update the test passes `expectedColor` directly after making the second draw using kStencilColor. Then, check if the result color is `expectedColor`. How is this approach?